### PR TITLE
[#1279] Provide a way to run analysis for a moving window

### PR DESCRIPTION
**Pull request by niqiukun**
Resolves https://github.com/reposense/RepoSense/issues/1279.

```
Some users might want to monitor repositories for a moving window rather
than from a specific date.

We can give a CLI option that runs the analysis for a period of specific 
length e.g., `--period 7d` runs the analysis for the last 7 days, and
`--period 1w` runs the analysis for the last 1 week.

Period can combine with either since date or until date, but not both.
If both not supplied, it will be the period before the current date.
```

### CODE REVIEW

**Comment by niqiukun** [comment]
UG to be updated after https://github.com/reposense/RepoSense/pull/1281 is merged.

**Comment by damithc** [comment]
@reposense/stage0-reviewers this PR needs reviews

File: docs/ug/cli.md [hunk 1]
```diff
@@ -88,6 +88,20 @@ This flag overrides the `Ignore standalone config` field in the CSV config file.
 
 <!-- ------------------------------------------------------------------------------------------------------ -->
 
+### `--period`, `-r`
```
Comment (0blivious) [inline docs/ug/cli.md:91]: -r should be changed to -p

File: docs/ug/cli.md [hunk 2]
```diff
@@ -88,6 +88,20 @@ This flag overrides the `Ignore standalone config` field in the CSV config file.
...
+
+**`--period, -p PERIOD`**: Specifies the period of analysis window.
```
Comment (0blivious) [inline docs/ug/cli.md:93]: remove the -p here?

**Comment by 0blivious** [comment]
May need to validate the calculated until date? Now if enter large period number the until date will be some day in future.

**Comment by niqiukun** [comment]
> May need to validate the calculated until date? Now if enter large period number the until date will be some day in future.

I think currently for `--until` we do not check if it is in future? Maybe we need to make the behaviour consistent.

**Comment by fzdy1914** [comment]
> In fact, there may be a use case where you set up the reposense report at the beginning of the semester and set the period to be 14 weeks (one semester). The until date can be someday in future so that you don't have to always update it as time progresses. I think that makes some sense.
 
@damithc Your input for this?

**Comment by damithc** [comment]
> > In fact, there may be a use case where you set up the reposense report at the beginning of the semester and set the period to be 14 weeks (one semester). The until date can be someday in future so that you don't have to always update it as time progresses. I think that makes some sense.
> 
> @damithc Your input for this?

Yes, it is good if we can cater for this use case. However, the report should be generated until _today_ or the _future end date_, whichever is earlier i.e., we should not show an empty space for the period from today to the future end date.

**Comment by niqiukun** [comment]
I have set the until date to be the earlier of the until date specified or current date. Not sure why `gradlew downloadNode` keeps failing though. Does not seem to be introduced by this PR.

**Comment by 0blivious** [comment]
> I have set the until date to be the earlier of the until date specified or current date. Not sure why `gradlew downloadNode` keeps failing though. Does not seem to be introduced by this PR.

CI failing

**Comment by niqiukun** [comment]
> CI failing

Any idea on the fix?

**Comment by 0blivious** [comment]
> Any idea on the fix?

@fzdy1914 @reposense/stage1-reviewers

**Comment by fzdy1914** [comment]
> > Any idea on the fix?
> 
> @fzdy1914 @reposense/stage1-reviewers

Seems the downloadNode gradle extension is outdated.

**Comment by fzdy1914** [comment]
> > > Any idea on the fix?
> > 
> > 
> > @fzdy1914 @reposense/stage1-reviewers
> 
> Seems the downloadNode gradle extension is outdated.

The issue happens in recent 2 days, and I am able to reproduce it with their latest plugins. Not sure if using java 11 will help.

**Comment by fzdy1914** [comment]
> > > > Any idea on the fix?
> > > 
> > > 
> > > @fzdy1914 @reposense/stage1-reviewers
> > 
> > 
> > Seems the downloadNode gradle extension is outdated.
> 
> The issue happens in recent 2 days, and I am able to reproduce it with their latest plugins. Not sure if using java 11 will help.

https://issues.liferay.com/browse/LPS-116536?jql=text%20~%20%22Node%20Gradle%20Plugin%3A%22 This is the related issue. It may be a good time to bump our Java version to 11 in the same time.

**Comment by fzdy1914** [comment]
@jamessspanggg need your review here

**Pull request merged**